trajopt/src/kinematic_terms.cpp

```diff
@@ -29,16 +29,49 @@ using Eigen::VectorXd;
 
 namespace trajopt
 {
+Eigen::VectorXd calculateExceededTolerance(const Eigen::VectorXd& lower_tolerance,
+                                           const Eigen::VectorXd& upper_tolerance,
+                                           const Eigen::VectorXd& error)
+{
+  // Initialize the resultant vector
+  Eigen::VectorXd resultant(error.size());
+
+  // Iterate through each element
+  for (int i = 0; i < error.size(); ++i)
+  {
+    // If error is within tolerances, set resultant to 0
+    if (error(i) >= lower_tolerance(i) && error(i) <= upper_tolerance(i))
+    {
+      resultant(i) = 0.0;
+    }
+    // If error is below lower tolerance, set resultant to error - lower_tolerance
+    else if (error(i) < lower_tolerance(i))
+    {
+      resultant(i) = error(i) - lower_tolerance(i);
+    }
+    // If error is above upper tolerance, set resultant to error - upper_tolerance
+    else if (error(i) > upper_tolerance(i))
+    {
+      resultant(i) = error(i) - upper_tolerance(i);
+    }
+  }
+
+  return resultant;
+}
+
```

Comment:
I think this already exists in Trajopt IFOPT in a vectorized, potentially faster way [here](https://github.com/tesseract-robotics/trajopt/blob/2d965b931015fba9a83ca697e114e3c07ef7c25b/trajopt_ifopt/src/utils/ifopt_utils.cpp#L119-L125)

Reply:
That comparison you referenced is doing something slightly different. I reimplemented this though in the latest commit.